File: oni/src/main/java/org/mbari/oni/jdbc/MutableConcept.java
/*
 * Copyright (c) Monterey Bay Aquarium Research Institute 2024
 *
 * oni code is non-public software. Unauthorized copying of this file,
 * via any medium is strictly prohibited. Proprietary and confidential. 
 */

package org.mbari.oni.jdbc;

import org.mbari.oni.etc.jdk.Logging;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class MutableConcept {
    private Long id;
    private MutableConcept parent;
    private String rank;
    private List<CName> names = new ArrayList<>();
    private List<MutableConcept> children = new ArrayList<>();
    private static final Logging log = new Logging(MutableConcept.class.getName());

    public MutableConcept() {
    }

    public MutableConcept(Long id, String rank, List<CName> names, MutableConcept parent, List<MutableConcept> children) {
        this.id = id;
        this.parent = parent;
        this.rank = rank;
        this.names = names;
        this.children = children;
    }

    public MutableConcept(Long id) {
        this.id = id;
    }

    public MutableConcept(Long id, String rank, List<CName> names, List<MutableConcept> children) {
        this.id = id;
        this.rank = rank;
        this.names = names;
        this.children = children;
    }

    public MutableConcept root() {
        if (parent == null) {
            return this;
        }
        return parent.root();
    }

    public MutableConcept copyUp() {
        return copyUp(List.of());
    }

    public MutableConcept copyUp(List<MutableConcept> newChildren) {
        var mc = new MutableConcept(id, rank, names, newChildren);
        mc.setParent(parent.copyUp(List.of(mc)));
        return mc;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getRank() {
        return rank;
    }

    public void setRank(String rank) {
        this.rank = rank;
    }

    public List<CName> getNames() {
        return names;
    }

    public void setNames(List<CName> names) {
        this.names = names;
    }

    public List<MutableConcept> getChildren() {
        return children;
    }

    public void setChildren(List<MutableConcept> children) {
        this.children = children;
    }

    public MutableConcept getParent() {
        return parent;
    }

    public void setParent(MutableConcept parent) {
        this.parent = parent;
    }

    public String getPrimaryName() {
        return names.stream()
                .filter(n -> n.nameType().equals("primary"))
                .findFirst()
                .map(CName::name)
                .orElse(null);
    }

    /**
     * Convert a list of ConceptRow objects to a tree of MutableConcept objects.
     * @param rows List of ConceptRow objects, basiclly the rows from the database table
     * @return A Pair of the root MutableConcept and a list of all MutableConcepts
     */
    public static TreeData toTree(List<ConceptRow> rows) {
        var nodes = new ArrayList<MutableConcept>();
        for (var row: rows) {

            // --- Find an existing parent MutableConcept or create one as needed.
            // This will be empty if it's the root node
            Optional<MutableConcept> parentOpt = Optional.empty();
            if (row.parentId() == null) {
                log.atInfo().log("Found root node: " + row);
            }
            else {
                parentOpt = Optional.of(nodes.stream()
                        .filter(n -> n.getId() != null && n.getId().equals(row.parentId()))
                        .findFirst()
                        .orElseGet(() -> {
                            var mc = new MutableConcept(row.parentId());
                            nodes.add(mc);
                            return mc;
                        }));
            }

            // --- Find the existing MutableConcept or create a new one
            MutableConcept self = nodes.stream()
                    .filter(n -> n.getId() != null && n.getId().equals(row.id()))
                    .findFirst()
                    .orElseGet(() -> {
                        var newMc = new MutableConcept(row.id());
                        nodes.add(newMc);
                        return newMc;
                    });

            parentOpt.ifPresent(parent -> {
                self.setParent(parent);
                parent.getChildren().add(self);
            });

            var cn = new CName(row.name(), row.nameType());
            self.setRank(row.rank().orElse(null));
            self.getNames().add(cn);
        }

        var root = nodes.stream()
                .filter(n -> n.getParent() == null)
                .findFirst()
                .orElseThrow(() -> new RuntimeException("No root node found"));
        return new TreeData(root, nodes);
    }

}

File: oni/src/main/scala/org/mbari/oni/jdbc/MutableConcept.scala
/*
 * Copyright 2024 Monterey Bay Aquarium Research Institute
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.mbari.oni.jdbc

import org.mbari.oni.domain.Concept

import scala.collection.mutable

class MutableConcept:
    var id: Option[Long]               = None
    var parent: Option[MutableConcept] = None
    var rank: Option[String]           = None
    var names: Seq[CName]              = Nil
    var children: Seq[MutableConcept]  = Nil

    def primaryName: Option[String] =
        names
            .find(n => n.isPrimary)
            .map(_.name)

    def alternativeNames: Seq[String] =
        names
            .filter(n => !n.isPrimary)
            .map(_.name)

    def copyUp(): MutableConcept = copyUp(Nil)

    private def copyUp(newChildren: Seq[MutableConcept]): MutableConcept =
        val mc = new MutableConcept
        mc.id = id
        mc.rank = rank
        mc.names = names
        mc.children = newChildren
        mc.parent = parent.map(_.copyUp(Seq(mc)))
        mc

    def toImmutable: Concept =
        // println(s"${this.id} - ${this.names}")
        val primaryName      = names
            .find(_.isPrimary)
            .getOrElse(names.head)
        val alternativeNames = names.filter(!_.eq(primaryName))
        Concept(
            primaryName.name,
            rank,
            alternativeNames.map(_.name),
            children.map(_.toImmutable)
        )

    def root(): MutableConcept = parent match
        case None    => this
        case Some(p) => p.root()

object MutableConcept:

    def newParent(parentId: Long): MutableConcept =
        val mc = new MutableConcept
        mc.id = Some(parentId)
        mc

    // TODO return the nodes too!!
    def toTree(rows: Seq[ConceptRow]): (Option[MutableConcept], Seq[MutableConcept]) =
        val nodes = new mutable.ArrayBuffer[MutableConcept]
        for row <- rows do

            /*
              Find an existing parent or create one as needed
             */
            val parentOpt = row
                .parentId
                .map(parentId =>
                    nodes
                        .find(_.id.getOrElse(-1L) == parentId)
                        .getOrElse {
                            val mc = newParent(parentId)
                            nodes += mc
                            mc
                        }
                )

            if parentOpt.isEmpty then
                System
                    .getLogger(getClass.getName)
                    .log(System.Logger.Level.DEBUG, s"No Parent found for $row")

            /*
              Find the existing concept or create one if needed
             */
            val concept = nodes.find(_.id.getOrElse(-1L) == row.id) match
                case None     =>
                    val mc = new MutableConcept
                    mc.id = Some(row.id)
                    nodes += mc
                    mc
                case Some(mc) => mc

            // Set the parent of the concept!!
            parentOpt.foreach(parent =>
                concept.parent = parentOpt
                if !parent.children.contains(concept) then parent.children = parent.children :+ concept
            )

            if concept.rank.isEmpty && row.rank.isDefined then concept.rank = row.rank

            val cn = CName(row.name, row.nameType)
            concept.names = concept.names :+ cn

        val root = nodes.find(_.parent.isEmpty)
        (root, nodes.toSeq)
